ignitionrobotics/ign-rendering
# Fix particle effect randomness

Signed-off-by: Ian Chen <ichen@osrfoundation.org>

# 🦟 Bug fix

## Summary

The randomness algorithm in the particle effect for has issue when `time` (which is used to add randomness) becomes large in the shaders. The fix is to just use the fractional part of the `time` value for adding randomness.

## To Test

To reproduce the problem, one way to test this is to artificially increase the `time` value in the shaders. You can do this by applying the patch below to this branch to simulate the old behavior when time is large:

```cpp
diff --git a/ogre2/src/media/materials/programs/depth_camera_fs.glsl b/ogre2/src/media/materials/programs/depth_camera_fs.glsl
index 0d3109d5..a22bd4b3 100644
--- a/ogre2/src/media/materials/programs/depth_camera_fs.glsl
+++ b/ogre2/src/media/materials/programs/depth_camera_fs.glsl
@@ -107,7 +107,7 @@ void main()
   if (particle.x > 0 && particleDepth < fDepth)
   {
     // apply scatter effect so that only some of the smoke pixels are visible
-    float r = rand(inPs.uv0 + vec2(rnd, rnd));
+    float r = rand(inPs.uv0 + vec2(rnd + 1e5, rnd + 1e5));
     if (r < particleScatterRatio)
     {
       // set point to 3d pos of particle pixel
```

Then run this example world:
https://gist.github.com/iche033/bcd3b7d3f4874e1e707e392d6dbb0aa0

The patch adds 100000 seconds (27.x hours) to time and you'll see bands across the rgbd camera :

![rgbd_particles_large_t](https://user-images.githubusercontent.com/4000684/130687326-b989613b-b63a-4491-addb-9bb71641bcdf.png)

The changes in this PR makes sure the random number never exceeds `1` and the noise pattern should stay distributed across the image:
![rgbd_particles](https://user-images.githubusercontent.com/4000684/130687580-9cc70da8-d7d7-4b36-a5da-b099790cdad4.png)



## Checklist

**Note to maintainers**: Remember to use **Squash-Merge**
